### ISSUE TITLE
Question about Varifocal loss

### ISSUE THREAD

**Issue by HAOCHENYE** [hyz-xmaster/VarifocalNet#8]
In the paper, the negtive weight of BCE loss is **alpha*p^gamma**. However, in varifocal_loss.py， the loss is implemented by:

focal_weight = target * (target > 0.0).float() + \
    alpha * **(pred_sigmoid - target)**.abs().pow(gamma) * \
    (target <= 0.0).float()

The negtive weight is **alpha*(p-q)^gamma**, why?

**Comment by hyz-xmaster** [comment]
This is the initial version of implementation of VFL and I forgot to refine it. 
`alpha * (pred_sigmoid - target).abs().pow(gamma) *
(target <= 0.0).float()`   actually equals to `alpha * pred_sigmoid.pow(gamma) *
(target == 0.0).float()`, because there is a multiplier `(target <= 0.0).float()` in that formula and the target is always >= 0.

**Comment by HAOCHENYE** [comment]
> This is the initial version of implementation of VFL and I forgot to refine it.
> `alpha * (pred_sigmoid - target).abs().pow(gamma) * (target <= 0.0).float()` actually equals to `alpha * pred_sigmoid.pow(gamma) * (target == 0.0).float()`, because there is a multiplier `(target <= 0.0).float()` in that formula and the target is always >= 0.

You means `alpha * pred_sigmoid.abs().pow(gamma) * (target <= 0.0).float()` equals   `alpha * pred_sigmoid.pow(gamma) * (target == 0.0).float()`  or   `alpha * (pred_sigmoid - target).abs().pow(gamma) * (target <= 0.0).float()`  equals to `alpha * pred_sigmoid.pow(gamma) * (target == 0.0).float()`?  I'd understand the situation if it is the former one.

According to paper, the negtive weight should be `alpha * pred_sigmoid.abs().pow(gamma) * (target <= 0.0).float()`.Is the formular of paper current version?

**Comment by hyz-xmaster** [comment]
Hi,  `target` is the IoU so it is always >= 0, which implies `target <= 0` <=> `target == 0`.  
In this way, 
`alpha * (pred_sigmoid - target).abs().pow(gamma) * (target <= 0.0).float()`  <=> 
`alpha * (pred_sigmoid - target).abs().pow(gamma) * (target == 0.0).float()`  <=> 
`alpha * pred_sigmoid.abs().pow(gamma) * (target == 0.0).float()`.

**Comment by HAOCHENYE** [comment]
Ohhh! Thanks, I understand it now.

**Comment by feiyuhuahuo** [comment]
![Screenshot from 2021-01-05 10-31-48](https://user-images.githubusercontent.com/32631344/103600058-86533280-4f41-11eb-8540-b7d56f6942b2.png)
![图片](https://user-images.githubusercontent.com/32631344/103600092-99fe9900-4f41-11eb-8bd1-9563b0cbb4b7.png)
Hi @hyz-xmaster ,
1. I did not find the `q` in the red circle according to the code.
2. I can't understand the item above the green line. Since `log(1-p)` is used to predict negative samples, why it appears in the `q>0` case? And anyway, I did not find the related implementation from the code. I just understand the code by the following way:
![Screenshot from 2021-01-05 10-41-02](https://user-images.githubusercontent.com/32631344/103600563-b3541500-4f42-11eb-831c-e464229d6d88.png)
 Looking forward to your reply, thanks.

**Comment by hyz-xmaster** [comment]
Hi @feiyuhuahuo,
1.  `target` in the code represents `q` in that formula. 
2.  `qlog(p)+(1-q)log(1-p)` is the binary cross entropy loss, which is calculated by [F.binary_cross_entropy_with_logits](https://github.com/hyz-xmaster/VarifocalNet/blob/5bbf1a6c29359c95cea1f5def8482f9aaa7365f8/mmdet/models/losses/varifocal_loss.py#L50). When `q  = 0`, `qlog(p)+(1-q)log(1-p)` reduces to `log(1-p)`. When `q > 0`, it keeps unchanged.

**Comment by yxx-byte** [comment]
> > This is the initial version of implementation of VFL and I forgot to refine it.
> > `alpha * (pred_sigmoid - target).abs().pow(gamma) * (target <= 0.0).float()` actually equals to `alpha * pred_sigmoid.pow(gamma) * (target == 0.0).float()`, because there is a multiplier `(target <= 0.0).float()` in that formula and the target is always >= 0.
> 
> You means `alpha * pred_sigmoid.abs().pow(gamma) * (target <= 0.0).float()` equals `alpha * pred_sigmoid.pow(gamma) * (target == 0.0).float()` or `alpha * (pred_sigmoid - target).abs().pow(gamma) * (target <= 0.0).float()` equals to `alpha * pred_sigmoid.pow(gamma) * (target == 0.0).float()`? I'd understand the situation if it is the former one.
> 
> According to paper, the negtive weight should be `alpha * pred_sigmoid.abs().pow(gamma) * (target <= 0.0).float()`.Is the formular of paper current version?

Hello, did you add your loss to yolov5? Judge which place needs to be adjusted?
<img width="406" alt="image" src="https://user-images.githubusercontent.com/73935151/170706827-16f5d347-9774-484e-9ad5-4bf42201b14f.png">
